# Variables

In programming, variables are like containers that hold data or values. They allow you to store and manipulate information in your program. Python is a dynamically typed language, which means you don't need to explicitly declare the type of a variable.

To create a variable in Python, you simply choose a name for it and assign it a value using the equal sign (`=`). Here's an example:

In [1]:
age = 25    # setting the integer value of 25 to the 
            # variable named `age`
print(age)  # printing the value of the variable `age`

25


In this example, `age` is the variable name, and `25` is the value assigned to it.

### Quick Assignment 1

Assign a positive numeric value to a variable, then print it.

---

## Variable Types

Python allows you to use different data types for variables, such as integers (whole numbers), floats (decimal numbers), strings (text), Boolean values (True or False), and more. The type of a variable is determined by the value it holds:

In [2]:
# an integer variable
age = 25
print(age)

25


In [3]:
# a float variable
height = 1.75
print(height)

1.75


In [4]:
# a string variable
first_name = "John"     # note: underscore is OK here.
print(first_name)

John


In [5]:
# a Boolean variable
is_student = True
print(is_student)

True


Variables can be updated by assigning them a new value at any point in your program. For example:

In [6]:
age = 25        # initial value
age = age + 1   # updating the value by adding 1
print(age)      # will output: 26

26


You can also perform operations on variables, combine them into expressions, and use their values in calculations. You can also reassign variables with the values of different type, example:

In [7]:
variable = 5
print(variable)
variable = "not a number"
print(variable)

5
not a number


However, this is not advised practice even if it works, because it will make your code harder to understand and maintain. However, sometimes it might be unavoidable.

### Quick Assignment 2

Assign a few variables of different types. Print their results. 

💡 Useful Note: `print()` function accepts multiple comma separated arguments

---

## Variable Reassignment

Variable can be reassigned to another variable. Here is the example:

In [8]:
number1 = 5
number2 = number1
print(number1, number2)
# both numbers will be 5

5 5


In [9]:
number1 = 10
print(number1, number2)
# while number1 will be set to 10, number2 will still be 5

10 5


### Quick Assignment 3

Set any value to a variable, then reassign it to another variable, and change the value of the old one. Print both variables before and after reassignment.

---

## Numeric Variable Types

In Python, integers (`int`) and floats (`float`) are two common numeric variable types. They differ in their representation and handling of numbers.

### Integer (`int`)

Integers are whole numbers without any decimal points. They can be positive or negative.

In [10]:
age = 25                # A positive integer
temperature = -10       # A negative integer
students_count = 100    # Another positive integer
print(age, temperature, students_count)

25 -10 100


In the above examples, `age` represents a person's age, `temperature` represents a negative temperature value, and `students_count` represents the number of students.

### Float (`float`)

Floats are numbers that have decimal points or fractional parts. They can also be positive or negative.

In [11]:
weight = 64.5               # A positive floating-point number
PI = 3.14159                # Another positive floating-point number
bank_balance = -2500.75     # A negative floating-point number
print(weight, PI, bank_balance)

64.5 3.14159 -2500.75


In this example, `weight` represents someone's weight in kilograms, `PI` represents the mathematical constant π (pi), and `bank_balance` represents a negative bank balance.

Integer variables can only represent whole numbers, while float variables offer greater precision and can represent fractional parts.

It's essential to choose the appropriate variable type (int or float) based on the nature of the number you are working with to ensure accurate calculations and proper representation of the data.

While floats are more flexible, operations with integers have much higher performance. Especially for iterations.

### Quick Assignment 4

Set a few integer and float variables and print them.

---

## Checking the type of a variable

To check the type of a variable in Python, you can use the built-in function `type()`. The `type()` function returns the type of an object or variable.

Here's how you can use it:

In [12]:
age = 25
print(type(age))        # Output: <class 'int'>

<class 'int'>


In [13]:
weight = 64.5
print(type(weight))     # Output: <class 'float'>

<class 'float'>


In [14]:
first_name = "John"
print(type(first_name))       # Output: <class 'str'>

<class 'str'>


In [15]:
is_student = True
print(type(is_student)) # Output: <class 'bool'>

<class 'bool'>


In this example, we have four variables of different types: `age` (integer), `weight` (float), `name` (string), and `is_student` (boolean). By applying the `type()` function to each variable and printing the result, we can determine the specific type of each variable.

The output of the `type()` function is displayed in the form of `<class 'typename'>`, where `'typename'` represents the specific type of the variable.

Using `type()` allows you to dynamically identify the type of a variable during runtime, which can be helpful during debugging or when working with code that involves different data types.

If you want a more human readable representation of the type object, you can use it's `__name__` property. Examples:

In [16]:
print(type(first_name).__name__)
print(type(is_student).__name__)

str
bool


### Quick Assignment 5

Print the types of variables defined in previous assignments. Also try to print pretty human readable names of such types.

---

## Variable Naming Rules and Exceptions

When naming a variable in Python, you need to adhere to certain rules and conventions. Here are the key guidelines for variable names:

1. Variable names must begin with a letter `a-z`, `A-Z` or an underscore `_` character. They cannot start with a number or special characters. Examples: `name`, `_count`, `result`. Variable names like `1st_place_winner` are invalid.

1. Variable names can contain letters (a-z, A-Z), numbers (0-9), and underscores (_). They are case-sensitive, so `age` and `Age` would be considered different variables. Examples of good variable names: `age`, `num_of_students`, `total_sum`, `player1`.

Here is also the example of how it is easy to make a mistake by not following a strict naming convention

In [17]:
age = 25
Age = 77
print(age, Age)

25 77


3. Reserved words, also known as keywords, are words that are part of the language's syntax. They have a specific meaning and purpose, and thus cannot be used as variable names. Examples include `if`, `for`, `while`, `print`, etc. By assigning variables to such reserved words will override the functions of these keywords and will cripple the programming language's base functionality.

1. Choose descriptive and meaningful names for variables to enhance code readability and understandability. For example, use `is_student` for boolean type variable.

1. Use lowercase letters and underscores for multi-word variable names to improve readability. This convention is known as snake_case. Example: `num_of_students`, `first_name`

1. While not required, it's common practice to use all uppercase letters for constants, variables whose values won't change throughout the program execution. Example: `PI = 3.14`, `DEFAULT_LANGUAGE = 'en-us'`

It's important to note that Python does not allow the use of special characters like `$` or `#` in variable names, and it is generally advisable to stick to the standard naming conventions to avoid confusion and maintain code consistency.

### Quick Assignment 6

Please explain why the code below returns False?

In [21]:
str = "some funny text"
print(type(str) == str)

False



### Additional tips

- Variable names should be self-explanatory to make the code more readable. For instance, instead of generic names like `x` or `temp`, choose more meaningful names like `counter` or `temperature`.

- It's a good practice to use plural nouns for variables representing collections. For instance, use `fruits` instead of `fruit` if it contains multiple fruits.

By following these naming guidelines, your code will become more understandable and maintainable, both for yourself and for others who read your code.

## Practical Assignment for recap

1. Create a variable and assign a number to it, then print out the variable's value

2. Add a negative number to the above variable and print the resulting value

3. Create a floating point variable, print it's value, then add it to the variables from the above tasks and print the results

4. Print out the type names of the variables created during the assignments 1-3.